File: backend/app/api/tasks.py
"""任务 CRUD + 统计 API"""
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.models import Task
from app.schemas import TaskCreate, TaskOut, TaskStatusUpdate, TaskUpdate

router = APIRouter(prefix="/api/tasks", tags=["tasks"])


@router.get("", response_model=list[TaskOut])
def list_tasks(
    search: str = Query("", description="搜索任务名称"),
    priority: str = Query("", description="按优先级筛选: high/medium/low"),
    sort: str = Query("deadline", description="排序字段: deadline/priority/created_at"),
    db: Session = Depends(get_db),
):
    q = db.query(Task)
    if search:
        q = q.filter(Task.task_name.contains(search))
    if priority:
        q = q.filter(Task.priority == priority)

    priority_order = {"high": 1, "medium": 2, "low": 3}
    q = q.order_by(
        Task.deadline.asc(),
        Task.id.desc(),
    )

    # 显式按优先级排序（Python 侧排序确保 high > medium > low）
    tasks = q.all()
    tasks.sort(key=lambda t: priority_order.get(t.priority, 99))

    # 补充邮件信息
    result = []
    for t in tasks:
        d = TaskOut.model_validate(t)
        if t.email:
            d.email_subject = t.email.subject
            d.email_sender = t.email.sender
        result.append(d)
    return result


@router.get("/stats")
def get_stats(db: Session = Depends(get_db)):
    return {
        "total": db.query(Task).count(),
        "high_priority": db.query(Task).filter(Task.priority == "high").count(),
        "active": db.query(Task).filter(Task.status != "done").count(),
        "done": db.query(Task).filter(Task.status == "done").count(),
    }


@router.get("/{task_id}", response_model=TaskOut)
def get_task(task_id: int, db: Session = Depends(get_db)):
    task = db.query(Task).get(task_id)
    if not task:
        raise HTTPException(404, "任务不存在")
    d = TaskOut.model_validate(task)
    if task.email:
        d.email_subject = task.email.subject
        d.email_sender = task.email.sender
    return d


@router.post("", response_model=TaskOut, status_code=201)
def create_task(body: TaskCreate, db: Session = Depends(get_db)):
    task = Task(
        task_name=body.task_name,
        deadline=body.deadline,
        priority=body.priority.value,
        status=body.status.value,
        category=body.category.value,
        confidence=1.0,
        confidence_source="manual",
    )
    db.add(task)
    db.commit()
    db.refresh(task)
    return TaskOut.model_validate(task)


@router.put("/{task_id}", response_model=TaskOut)
def update_task(task_id: int, body: TaskUpdate, db: Session = Depends(get_db)):
    task = db.query(Task).get(task_id)
    if not task:
        raise HTTPException(404, "任务不存在")
    for field in ("task_name", "deadline", "priority", "category", "status"):
        val = getattr(body, field, None)
        if val is not None:
            setattr(task, field, val.value if hasattr(val, "value") else val)
    db.commit()
    db.refresh(task)
    return TaskOut.model_validate(task)


@router.delete("/{task_id}", status_code=204)
def delete_task(task_id: int, db: Session = Depends(get_db)):
    task = db.query(Task).get(task_id)
    if not task:
        raise HTTPException(404, "任务不存在")
    db.delete(task)
    db.commit()


@router.patch("/{task_id}/status", response_model=TaskOut)
def update_status(task_id: int, body: TaskStatusUpdate, db: Session = Depends(get_db)):
    task = db.query(Task).get(task_id)
    if not task:
        raise HTTPException(404, "任务不存在")
    task.status = body.status.value
    db.commit()
    db.refresh(task)
    return TaskOut.model_validate(task)

File: frontend/src/stores/tasks.ts
import { ref, computed } from 'vue'
import { defineStore } from 'pinia'
import api from '../api'

export interface Task {
  id: number
  email_id: number | null
  task_name: string
  deadline: string
  priority: 'high' | 'medium' | 'low'
  status: 'pending' | 'processing' | 'done'
  category: string
  confidence: number
  confidence_source: string
  created_at: string
  email_subject?: string | null
  email_sender?: string | null
}

export interface Stats {
  total: number
  high_priority: number
  active: number
  done: number
}

export const useTaskStore = defineStore('tasks', () => {
  const tasks = ref<Task[]>([])
  const stats = ref<Stats>({ total: 0, high_priority: 0, active: 0, done: 0 })
  const loading = ref(false)
  const search = ref('')
  const priorityFilter = ref('')

  const pendingTasks = computed(() => tasks.value.filter(t => t.status === 'pending'))
  const processingTasks = computed(() => tasks.value.filter(t => t.status === 'processing'))
  const doneTasks = computed(() => tasks.value.filter(t => t.status === 'done'))

  async function fetchTasks() {
    loading.value = true
    try {
      const params: Record<string, string> = {}
      if (search.value) params.search = search.value
      if (priorityFilter.value) params.priority = priorityFilter.value
      const res = await api.get('/tasks', { params })
      tasks.value = res.data
    } finally {
      loading.value = false
    }
  }

  async function fetchStats() {
    const res = await api.get('/tasks/stats')
    stats.value = res.data
  }

  async function createTask(data: {
    task_name: string
    deadline: string
    priority: string
    category: string
    status: string
  }) {
    const res = await api.post('/tasks', data)
    tasks.value.unshift(res.data)
    await fetchStats()
    return res.data
  }

  async function updateTask(id: number, data: Record<string, unknown>) {
    const res = await api.put(`/tasks/${id}`, data)
    const idx = tasks.value.findIndex(t => t.id === id)
    if (idx >= 0) tasks.value[idx] = res.data
    return res.data
  }

  async function updateStatus(id: number, status: string) {
    // 乐观更新
    const idx = tasks.value.findIndex(t => t.id === id)
    if (idx >= 0) tasks.value[idx].status = status as Task['status']

    try {
      const res = await api.patch(`/tasks/${id}/status`, { status })
      if (idx >= 0) tasks.value[idx] = res.data
      await fetchStats()
    } catch {
      // 回滚
      await fetchTasks()
      await fetchStats()
    }
  }

  async function deleteTask(id: number) {
    await api.delete(`/tasks/${id}`)
    tasks.value = tasks.value.filter(t => t.id !== id)
    await fetchStats()
  }

  return {
    tasks, stats, loading, search, priorityFilter,
    pendingTasks, processingTasks, doneTasks,
    fetchTasks, fetchStats, createTask, updateTask, updateStatus, deleteTask,
  }
})
